### PR TITLE
WIP: game logic rewrite

### PR DESCRIPTION
**Pull request by Leptopoda**
continuation of #417 

This PR aims to rewrite the game logic (in lib/mill and game_page.dart).
It is still WIP as the issue mentioned in #417 still exists. Everything else should work though. 

The merge conflict is connected to #416 as I found a more elegant solution.

known problems:
- [x] step forward all somehow shows a bad snackbar
- [x] ai is broken :) (see #417) [sort of fixed]
- [x] ai hangs

### CODE REVIEW

File: src/ui/flutter_app/lib/mill/src/types.dart [hunk 1]
```diff
@@ -157,6 +158,22 @@ extension PieceColorExtension on PieceColor {
         return S.of(context).draw;
       case PieceColor.ban:
       case PieceColor.nobody:
+        throw Exception("Player has no name");
```
Comment (calcitem) [inline src/ui/flutter_app/lib/mill/src/types.dart:58]: At least when the value is `ban`, an exception should not be thrown.

File: src/ui/flutter_app/lib/mill/src/types.dart [hunk 2]
```diff
@@ -224,7 +224,11 @@ extension PieceColorExtension on PieceColor {
     }
   }
 
-  IconData get chevron {
+  IconData get icon {
```
[the remark this answers is not in the document]
Comment (Leptopoda) [reply src/ui/flutter_app/lib/mill/src/types.dart:124]: But i think they are related to the class.
Like the icon a PieceColor object represents...

File: src/ui/flutter_app/lib/mill/src/types.dart (hunk 1 above)
Comment (Leptopoda) [reply src/ui/flutter_app/lib/mill/src/types.dart:58]: Iirc there was an `assert(false)` in exactly the same place. 

In this case maybe the string is nit that accurate but it still applies. 

As the exceptions are not runtime related we should probably change them to Errors anyways

**Comment by Leptopoda** [comment]
hi there @calcitem 

small update :upside_down_face:  the hanging of the ai has been fixed.
FYI the bugs weren't logic bugs (good for me xD) but they where:
- a typo in a FEN string (somehow a comma slipped in)
- a swapped if statement ( != null instead of == null)
- an enum that was wrongly converted to FEN string

You can now play a bit around with the changes but no need for a propper review yet. There is still a lot to come!

**Comment by calcitem** [comment]
- [ ] If the game does not automatically restart automatically after the modification rule settings, the AI exception may be triggered.

**Comment by calcitem** [comment]
I am making a new Git commit to resolve conflicts with branch dev.

**Comment by calcitem** [comment]
The synchronization with the branch dev has been done, the build has passed, and I will not continue to change this branch now.

**Comment by Leptopoda** [comment]
> The synchronization with the branch dev has been done, the build has passed, and I will not continue to change this branch now.

Thanks for the work but I'll probably override them by force pushing. I already have quite a few changes locally...

I think the big changes should be done soon so we can test and commit the changes. I'll also make a list of thing to closely look at. All the other todos I added can then be resolved separately.

**Comment by calcitem** [comment]
No problem, you can force it to be overwritten, as long as you don't overwrite the changes in the dev branch.  Looking forward to more improvements!

**Comment by Leptopoda** [comment]
Ok I think this one is ready for review now. 

Pay close attention to:
- directionality (game nav bar and other icons) see #408 

Known issues:
- derectionality of the winning icon
- default algorithm

> After the devmode removal, the dynamic depth test is now not easily performed, so the test efficiency is reduced. This is the point I ignored before. Just make a record first, follow-up additions to the detailed description. The fastest way is to let CI can build a DevMode apk.

what do you mean by that?

**Comment by calcitem** [comment]
- [ ] The pieces that have just moved recently will become very large.

- [ ] AI Vs AI is hang.

**Comment by Leptopoda** [comment]
>  The pieces that have just moved recently will become very large.

rip I somehow forgot to push the fix for that one...

>  AI Vs AI is always hang.

that'll need some work xD

>  Sometimes when start a new game, and tap, AI is hang. Showing "Thinking..." and no response.

Not related to my changes but I'll fix that. It might alter the UI a bit.
The problem is that the engine needs to be started. You did this in `initState()` but as this call isn't asynchronous it won't await the future of `_engine.startup()`. This wasn't a problem before as there were a lot of other tasks that needed to be done thus the slow initialization covered this up. I'm going to use a `FutureBuiler` now but we might need to think of a good way to display that it's loading.

>  Sometimes when exporting move list, the move list cannot copied to clipboard.

I need logs for this one. This probably is a bug outside of our scope.

>  Export and Import game. Get message: Moves and game rules do not match. See movelist, is broken

Can you actually tell me how the move list should have been? I might need that in case I can't reproduice it (last time I ex/imported games it worked)

>  When tap info button, exception:

:upside_down_face: happens I guess...

**Comment by Leptopoda** [comment]
Ok I marked to ones I have resolved locally.
As said above I need more Information on a few things. Also I can't reproduce the bad/duplicate export. How did you manage to break it that way :upside_down_face:

The import and export services work fine :woman_shrugging:

**Comment by calcitem** [comment]
# Move list is broken

Branch: cleanup/enums
Commit 689d9ff3140c72c40e3e1e44677661a03408ebe1

IDE: Android Studio

Run `./flutter-windows-init.sh`

Delete `Documents/Sanmill`

Click `Debug main.dart`

Tap `Open navigation menu -> Preferences`, and set:

```
Move randomly: off

AI thinking time: 60
```

Make sure the log shows:

`option name Shuffling type check default false`

(Sometimes it shows `true`,  I don't know why. Maybe it is another bug)

Human taps `d2`, AI may tap `d6`.

Tap `Game -> New game`

Human taps `d2`, AI may tap `f4`. (Not same as `d6`, it is unexpected)

Tap `Move -> Move list`

It shows:

```
Move list
1. d2 1. d6
2. d2 2. f4
```

Expected:

```
Move list
1. d2 f4
```

This bug can be reproduced on Android phones.

**Comment by calcitem** [comment]
# LateInitializationError: Field 'blurPositionColor' has not been initialized.

Set rules:

```
General
  The number of pieces each player has : 12
  Diagonal lines : ON
  
Playcing
  Mark and delay removing pieces : ON

Moving
  The second player moves first : ON

Flying
  Flying : OFF

Removing
  Destroy mills : ON
```

Human Vs. AI.

When tapping the board. Exception occurs:

```
LateInitializationError: Field 'blurPositionColor' has not been initialized.
```

**Comment by calcitem** [comment]
# Piece style cannot be adjusted back to "none" again

Personalization -> Point style -> Solid
Personalization -> Point style -> None

Can't take effect.

**Comment by calcitem** [comment]
# The screen reader cannot read board coordinates. 

Commit 50aa235 affects performance. Another approach needs to solve this functional problem.

**Replication steps**

1. Preferences -> Accessibility -> Screen reader support : ON
2. Turn on TalkBack.
3. Tap the board.
4. Problem: board coordinates cannot be read by TalkBack.

**Comment by calcitem** [comment]
# The unplaced piece counter in the center of the board is not displayed. It is only displayed when placed piece.

**Replication steps**

1.  Personalization -> Display -> Show count of unplaced pieces : ON
2. Return to the game page
3. No piece counter shown in the center of the board
4. Only when placing the first piece, the counter is shown.

I don't know if this is an intentional design change or an unexpected bug.

**Comment by calcitem** [comment]
> After the dev mode removal, the dynamic depth test is not easily performed, so the test efficiency is reduced. This is the point I ignored before. Just make a record first, follow-up additions to the detailed description. The fastest way is to let CI can build a DevMode apk.

The AI's search depth is fixed in the developer mode, and the "randomly move" switch is disabled, so all the operation steps can be copied in full without random changes. So I need to switch to this mode to test the app, but I haven't figured out how to change to dev mode yet. It would be nice to have a web page detailing it.

**Comment by Leptopoda** [comment]
> # The screen reader cannot read board coordinates.
> 
> Commit [50aa235](https://github.com/calcitem/Sanmill/commit/50aa23521d99a6d7e9f25fd0920b24a38060192c) affects performance. Another approach needs to solve this functional problem.
> 
> **Replication steps**
> 
>     1. Preferences -> Accessibility -> Screen reader support : ON
> 
>     2. Turn on TalkBack.
> 
>     3. Tap the board.
> 
>     4. Problem: board coordinates cannot be read by TalkBack.

Just FYI there is no genaral performance problem but we shouldn't render the overlay when it isn't used.

```dart
      child: EnvironmentConfig.devMode ? const _DevGrid() : null,
```

should probably be 

```dart
      child: (EnvironmentConfig.devMode || _prefs.screenReaderSupport)? const _DevGrid() : null,
```

In this case the _DevGrid() will also be renamed :upside_down_face:

**Comment by Leptopoda** [comment]
>>    For ease of use, some launch configs for IntelliJ IDEA or Visual Studio Code are available. Just select the needed one in the Run and Debug or Run/Debug Configurations tab.
>
>I cannot find it. Need I create it?
Screenshot_20211212_114331
Screenshot_20211212_114418

**Comment by calcitem** [comment]
I don't know whether these configuration items in IntelliJ IDEA are loaded from the `.run` directory or created manually. I didn't find this drop-down list in my Android Studio, only `MAIN.DART`. Maybe Android Studio and IntelliJ IDEA use different configuration directories. I will try to create the configuration myself.

**Comment by calcitem** [comment]
@all-contributors please add @Leptopoda  for code, doc, ideas,  maintenance, projectManagement, platform,  review, and  translation

**Comment by calcitem** [comment]
@all-contributors please add @Leptopoda  for code, doc, ideas,  maintenance, projectManagement, platform,  review, and  translation

**Comment by calcitem** [comment]
@all-contributors please add @Leptopoda  for code, doc, ideas,  maintenance, projectManagement, platform,  review, and  translation

**Comment by Leptopoda** [comment]
in the tab edit

**Comment by calcitem** [comment]
I finally know why I can't display the configurations because I opened the `flutter_app` directory with the IntelliJ IDEA. The configurations can be loaded generally after copying `.run` to the `flutter_app` directory.

I am curious why you open the root directory directly with the IntelliJ IDEA? Generally speaking, you do not need to modify files outside the `flutter_app` directory. There is an inconvenience in opening the root directory: you may accidentally replace the text in the C++ code when replacing text.

**Comment by Leptopoda** [comment]
>  The unplaced piece counter in the center of the board is not displayed. It is only displayed when placed piece.

can't reproduce this one

> Sometimes when exporting move list, the move list cannot be copied to the clipboard.

> Export and Import game. Get message: Moves and game rules do not match. See move list, is broken

As already said those probably originate from a device specific bug...

>  Same log repeat three times

reminds me of your broken game history... reproduction steps are welcome

TODO:

> The pieces that have just moved recently will become very large.

> AI Vs AI is always hung.

> Move list is broken

> LateInitializationError: Field 'blurPositionColor' has not been initialized.

<s> >  LateInitializationError: Field 'blurPositionColor' has not been initialized.

did this happen after a hot reload?
Can't reproduce that either. `blurPositionColor` is only set in the constructor and it only gets called once... </s>

**Comment by Leptopoda** [comment]
if the above mentioned issues are the only ones I'd like to merge the PR anyways please. 
- The late initialization errors are all not production critical and going to be fixed later...
- the aivsai bug is just that the game doesn't start the game automatically. You currently need to manually press "new game".

**Comment by calcitem** [comment]
https://github.com/calcitem/Sanmill/pull/421#issuecomment-988205017

Can you tick the problem that has been fixed?  I can check again.  The remaining problems can be fixed later, and some of the problems that cannot be reproduced will be shown by recording gif.

**Comment by Leptopoda** [comment]
I already ticked the relevant ones. The ones not ticked are either for the next pr or for not reproducible as mentioned above

**Comment by calcitem** [comment]
AI Vs. AI, first game is OK, second game is hang.

**Comment by calcitem** [comment]
OK I will verify them in 3 hours.

**Comment by calcitem** [comment]
LGTM. For the remaining bugs, I will open Issues to follow up.

**Pull request merged**